rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricSummary.java

```diff
@@ -0,0 +1,505 @@
+        public @Nullable InterruptibleFuture<List<Either<String, MarkedString>>> getDocumentation(Position cursor) {
+            return get(documentation, cursor);
+        }
+
+        @Override
+        public @Nullable InterruptibleFuture<List<Location>> getDefinitions(Position cursor) {
+            return get(definitions, cursor);
+        }
+
+        @Override
+        public @Nullable InterruptibleFuture<List<Location>> getReferences(Position cursor) {
+            return get(references, cursor);
+        }
+
+        @Override
+        public @Nullable InterruptibleFuture<List<Location>> getImplementations(Position cursor) {
+            return get(implementations, cursor);
+        }
+
+        @Override
+        public void invalidate() {
+            super.invalidate();
+            documentation.interrupt();
+            definitions.interrupt();
+            references.interrupt();
+            implementations.interrupt();
+        }
+
+        private <T> InterruptibleFuture<Lazy<IRangeMap<List<T>>>> mapCalculation(String logName,
+                InterruptibleFuture<IConstructor> calculation, String kwField, Function<IValue, T> valueMapper) {
+
+            logger.trace("{}: Mapping summary by getting {}", logName, kwField);
+            return calculation
+                .thenApply(IConstructor::asWithKeywordParameters)
+                .thenApply(s ->
+                    Lazy.defer(() ->
+                        translateRelation(logName, getKWFieldSet(s, kwField), valueMapper)));
+        }
+
+        private IRelation<ISet> getKWFieldSet(IWithKeywordParameters<? extends IConstructor> data, String name) {
+            if (data.hasParameter(name)) {
+                return ((ISet) data.getParameter(name)).asRelation();
+            }
+            return IRascalValueFactory.getInstance().set().asRelation();
+        }
+
+        private <T> IRangeMap<List<T>> translateRelation(String logName,
+                IRelation<ISet> binaryRel, Function<IValue, T> mapValue) {
+
+            logger.trace("{}: summary contain rel of size:{}", logName, binaryRel.asContainer().size());
+            TreeMapLookup<List<T>> result = new TreeMapLookup<>();
+            for (IValue v: binaryRel) {
+                ITuple row = (ITuple)v;
+                Range from = Locations.toRange((ISourceLocation)row.get(0), columns);
+                T to = mapValue.apply(row.get(1));
+                var existing = result.getExact(from);
+                if (existing == null) {
+                    // most cases there is only a single entry, to so save a lot of memory, we store a singleton list to start with
+                    result.put(from, Collections.singletonList(to));
+                }
+                else if (existing.size() == 1) {
+                    // we had a singleton list in there, so let's replace it with a regular list
+                    existing = new ArrayList<>(existing);
```

Comment:
Use Collections.singletonList?

Reply:
We did, this is the path to upgrade to a normal list.